File: solutions/c/grains/1/grains.c
#include "grains.h"

uint64_t square(uint8_t index) {
    if (index == 1) {
        return 1;
    } else {
        return 2 * square(index-1);
    }
}
uint64_t total(void) {
    uint64_t ans = 0;
    uint64_t sq = 1;
    for (int i = 0; i < 64; i++) {
        ans += sq;
        sq *= 2;
    }
    return ans;
}

File: solutions/zig/grains/1/grains.zig
pub const ChessboardError = error{IndexOutOfBounds};

pub fn square(index: usize) ChessboardError!u64 {
    if (index == 0 or index > 64) {
        return ChessboardError.IndexOutOfBounds;
    } else {
        const b: usize = 1;
        const p: u6 = @intCast(index - 1);
        const ans: usize = b << p;
        return ans;
    }
}

pub fn total() u64 {
    var ans: u64 = 0;
    for (1..65) |i| {
        const b: usize = 1;
        const p: u6 = @intCast(i - 1);
        ans += b << p;
    }
    return ans;
}
